pkgs/dart_services/lib/src/common_server.dart

```diff
@@ -200,6 +202,54 @@ class CommonServerApi {
     return ok(version().toJson());
   }
 
+  static String? googleApiKey = Platform.environment['GOOGLE_API_KEY'];
+  http.Client? geminiHttpClient;
+
+  @Route.post('$apiPrefix/_gemini')
+  Future<Response> gemini(Request request, String apiVersion) async {
+    if (apiVersion != api3) return unhandledVersion(apiVersion);
+
+    // Read the api key from env variables (populated on the server).
+    final apiKey = googleApiKey;
+    if (apiKey == null) {
+      return Response.internalServerError(
+          body: 'gemini key not configured on server');
+    }
+
+    // Only allow the call from known clients / endpoints.
+    const firebaseHostAddress = '199.36.158.100';
```

Comment:
Would this be the address used if calling from a Firebase-hosted cloud function or similar?

Reply:
An update here - after an in-person discussion w/ @johnpryan, I switched to checking that the request origin is `https://dartpad.dev`.